Add USDC grants rejection system

damianmarti opened the pull request:
Added a rejection system for USDC grants, which needs 2 rejections to reject a grant or a new stage.

The first rejection shows a confirmation dialog:

![localhost_3000_admin (12)](https://github.com/user-attachments/assets/40ef32cc-04b1-46d0-a4b4-d53cbc041075)

The second one shows who previously rejected the grant or new stage, adds a clear message that the status will change to **Rejected**, and allows adding a note to the grantee.

![localhost_3000_admin (11)](https://github.com/user-attachments/assets/24aa18ec-e7ba-44cd-94eb-f94ebf7ec063)

The pre-approvals and pre-rejects are showing thumbs up or down on the proposal admin card.

![localhost_3000_admin (13)](https://github.com/user-attachments/assets/73958254-d626-4326-b02b-826163df8210)

related to #62

damianmarti commented on packages/nextjs/app/api/large-stages/[stageId]/reject-vote/route.ts:
```diff
@@ -0,0 +1,59 @@
+    });
+
+    const grant = await getStageByIdWithGrantAndVotes(Number(stageId));
+    if (grant && grant.rejectVotes.length === MINIMAL_VOTES_FOR_FINAL_APPROVAL) {
```
Used the same constant from approvals since it makes sense that they have the same value.

damianmarti commented on packages/nextjs/app/admin/_components/LargeGrantProposal.tsx:
```diff
@@ -93,6 +102,20 @@ export const LargeGrantProposal = ({ proposal, userSubmissionsAmount, isGrant }:
+        {(latestStage.approvalVotes.length > 0 || latestStage.rejectVotes.length > 0) && (
+          <div className="flex gap-1 justify-end">
+            {latestStage.approvalVotes.map(vote => (
+              <div key={vote.id} className="tooltip" data-tip={`Pre-approved by ${vote.authorAddress}`}>
```
This is showing the address without using the Address component because we need to update to DaisyUI v5 to be able to use tooltip-content https://daisyui.com/docs/v5/

We can improve this later (without updating to DaisyUI v5), resolving the authorAddresses and showing the ENS names (if they have).

rin-st replied:
Create an issue please so we don't forget

Pabl0cks reviewed (commented):
Nice job Damu! I really like this new flow. A few comments after the first round of functional test:

- When you vote for rejection, the "Already rejected" label makes the button section look a bit weird, changing to "Already voted" makes it consistant with approvals and it looks nicely styled:


| Rejected| Voted|
|:---:|:---:|
| <img src="https://github.com/user-attachments/assets/f97ed550-cca3-4921-953c-55832e14a1eb" width="400"> | <img src="https://github.com/user-attachments/assets/4dfffbe7-72b4-4b0f-9842-9034257e51fa" width="400"> |
- When you do the final rejection adding a rejection note, i get this error: "Recovered address did not match session address", if I delete the note, works fine.
(Sorry for my night filters 😂)

![image](https://github.com/user-attachments/assets/515c153d-00b6-47db-bd2e-dea6ae4e1287)
- Would be nice to force these error to show on top of the modal grey background adding the `<Toaster />` before the `</dialog>`

- Nitpick: the initial rejection modal maybe should have the button copy as "Vote for rejection". Can totally ignore

![image](https://github.com/user-attachments/assets/dcc2a678-d640-4318-82e9-0e5b3c2a576c)


Will finish testing tomorrow 🙏

damianmarti commented:
> Nice job Damu! I really like this new flow. A few comments after the first round of functional test:

Thanks for testing!!

> 
> * When you vote for rejection, the "Already rejected" label makes the button section look a bit weird, changing to "Already voted" makes it consistant with approvals and it looks nicely styled:
> 
> Rejected	Voted
> <img alt="" width="400" src="https://private-user-images.githubusercontent.com/55535804/434080177-f97ed550-cca3-4921-953c-55832e14a1eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjY3NDksIm5iZiI6MTc0NDgyNjQ0OSwicGF0aCI6Ii81NTUzNTgwNC80MzQwODAxNzctZjk3ZWQ1NTAtY2NhMy00OTIxLTk1M2MtNTU4MzJlMTRhMWViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE4MDA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmMjcyY2ZiMzRkMzgzODcwNmQ2YWJkNTQyODFlMDc3ZThiYzQ4ZjE3N2E3ZDI4MDJhMTc3ODBkNjg5OTVhZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.supsfM6NjwyUxc2CoLfVNPRHWMI0y851F2UZw5c5WT4">	<img alt="" width="400" src="https://private-user-images.githubusercontent.com/55535804/434080165-4dfffbe7-72b4-4b0f-9842-9034257e51fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjY3NDksIm5iZiI6MTc0NDgyNjQ0OSwicGF0aCI6Ii81NTUzNTgwNC80MzQwODAxNjUtNGRmZmZiZTctNzJiNC00YjBmLTk4NDItOTAzNDI1N2U1MWZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE4MDA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0Y2M0YWNhOWM1YzI2ZjQ4YWI5NDkwN2I0MzdjYWU0MWM5NWE3ZDg2YWIxYWE4ZjBiNDQyMTZiMTA5Y2UyMDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h6QHjM9dMTwovmP9yW3k9bq6td2dBpZdW-ZPul4Cruk">


It was shown fine to me, but I changed the text anyway.

![localhost_3000_admin (14)](https://github.com/user-attachments/assets/ba9f5b3b-593c-449f-b02a-91a311f55d68)

> * When you do the final rejection adding a rejection note, i get this error: "Recovered address did not match session address", if I delete the note, works fine.
>   (Sorry for my night filters 😂)
> 
> ![image](https://private-user-images.githubusercontent.com/55535804/434080950-515c153d-00b6-47db-bd2e-dea6ae4e1287.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjY3NDksIm5iZiI6MTc0NDgyNjQ0OSwicGF0aCI6Ii81NTUzNTgwNC80MzQwODA5NTAtNTE1YzE1M2QtMDBiNi00N2RiLWJkMmUtZGVhNmFlNGUxMjg3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE4MDA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5NjRkZjVlZmEwZTYyYWJmZjQ5NzM0ZTIzYjdjMDQ0ODhiYTE5YjVkNjE3MTI1NGFmMDkyZDA0NTBkNDhmN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1gmD5vfKWsg92qFh-hiRBYRmyEhBX2H-D0H2xYyY1GQ)
> 

Fixed!

> * Would be nice to force these error to show on top of the modal grey background adding the `<Toaster />` before the `</dialog>`

Done!

> * Nitpick: the initial rejection modal maybe should have the button copy as "Vote for rejection". Can totally ignore
> 
> ![image](https://private-user-images.githubusercontent.com/55535804/434082319-dcc2a678-d640-4318-82e9-0e5b3c2a576c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjY3NDksIm5iZiI6MTc0NDgyNjQ0OSwicGF0aCI6Ii81NTUzNTgwNC80MzQwODIzMTktZGNjMmE2NzgtZDY0MC00MzE4LTgyZTktMGU1YjNjMmE1NzZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE4MDA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0ZjkyOThkYzBmMWJlYWNkOGFlYjdlN2YzOTExYjM5Y2Q4NDVhMjE1MGIzMTljOTQ2ODhiMzJkMGNkOGYwNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Qt2EdySUCjBFnNVZ6yYhG1mw3bJfbf7c_xl1z-kJTBQ)
> 

Changed!

> Will finish testing tomorrow 🙏

Thanks!!

Pabl0cks reviewed (approved):
Thanks for the quick fixes Damu! 

LGTM on the functional side, @rin-st could you review the code when you can? 🙏

rin-st reviewed (commented):
Looks good, thanks! Some comments/nitpicks

Bug: I can vote for both rejection and approval from the same account. It should be restricted

---

ui:

<img width="534" alt="image" src="https://github.com/user-attachments/assets/aeabf6c4-bb38-48e2-a144-0fd824553cd2" />

- We have `Cancel` button for large stages and large grants but don't have for milestones and for regular grants/stages. Probably it's easier just remove that button at all
- Modal buttons has different sizes and paddings across all page modals. For example, `Cancel` button is smaller than `Reject`. 48px vs 52px height. I think they should be the same. Probably other PR
<img width="558" alt="image" src="https://github.com/user-attachments/assets/7ef5e14b-934e-4857-b5a5-bb41f40cce92" />
<img width="491" alt="Screenshot 2025-04-17 at 10 31 19" src="https://github.com/user-attachments/assets/ab6cb6f2-218c-4e6c-9076-c20f9cbe1095" />


- `Reject` button is always Red. I think we can make it lighter when it's not final reject similar to approving
- It would be good to change text from `Reject` to `Vote for rejection` but we don't have space for it :(. Update: we have that button inside large stage reject modal. Maybe it's better to add it also for grants/milestones
- Text: "Clicking Reject will change the status of this grant to Rejected". It's not right. Status will change after final reject but not after click. As I remember purpose of that text was to warn about one-click rejection. Since reject behavior is changed for large grants probably it's better just to make text as for approval, something like "Are you sure you want to vote for the rejection of this stage?". Update: it already fixed but only for large stages
- Not related to thihs PR but we have this text also for regular (eth) **stages**: "Clicking Reject will change the status of this grant to Rejected". Need to change "grant" to "stage"

> When you vote for rejection, the "Already rejected" label makes the button section look a bit weird, changing to "Already voted" makes it consistant with approvals and it looks nicely styled:

note: for me, it's still weird 🤷‍♂️ . Mac 16 inches, browser window is full monitor width, scaling is 100%. 
<img width="590" alt="image" src="https://github.com/user-attachments/assets/3e56ba86-1c8e-42f3-8f8b-944ebdb93360" />

Pabl0cks commented:
> Bug: I can vote for both rejection and approval from the same account. It should be restricted

That one felt a bit strange to me too, but we were already letting them reject on ETH grants after an approval vote. 
I think maybe it’s OK to leave it as is for both USDC and ETH grants, since they’re admins and not regular users.

damianmarti commented:
> Looks good, thanks! Some comments/nitpicks

Thanks for the review!!

> 
> Bug: I can vote for both rejection and approval from the same account. It should be restricted

Yeah, like Pablo said, it's used only by admins, I think it's not a big issue. Anyway, if we want to restrict it, it's an easy fix, at least from the frontend.

> 
> ui:
> 
> * We have `Cancel` button for large stages and large grants but don't have for milestones and for regular grants/stages. Probably it's easier just remove that button at all

The Cancel button was added to the modal that are only a confirmation. I think this button makes sense there, because it feel strange that you don't have a cancel button on a confirmation dialog. Maybe we can add the Cancel button on all modals too.

> * Modal buttons has different sizes and paddings across all page modals. For example, `Cancel` button is smaller than `Reject`. 48px vs 52px height. I think they should be the same. Probably other PR

Weird, I will take a look.

> 
> <img alt="image" width="558" src="https://private-user-images.githubusercontent.com/25638585/434687716-7ef5e14b-934e-4857-b5a5-bb41f40cce92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDA5NjEsIm5iZiI6MTc0NTI0MDY2MSwicGF0aCI6Ii8yNTYzODU4NS80MzQ2ODc3MTYtN2VmNWUxNGItOTM0ZS00ODU3LWI1YTUtYmI0MWY0MGNjZTkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDEzMDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2MmU0YjY2OTEzNTM0OGQ3Y2U5YjdkYzM2MmNlN2RmNTVmOWY3OTM3ZDNiMzJiMGFkZDc3MzhiNjkzOGQ3MTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IgeJSSubKbD14tqVYyLTzMdU250yltVjDsnLjcC1FJ0"> <img alt="Screenshot 2025-04-17 at 10 31 19" width="491" src="https://private-user-images.githubusercontent.com/25638585/434693721-ab6cb6f2-218c-4e6c-9076-c20f9cbe1095.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDA5NjEsIm5iZiI6MTc0NTI0MDY2MSwicGF0aCI6Ii8yNTYzODU4NS80MzQ2OTM3MjEtYWI2Y2I2ZjItMjE4Yy00ZTZjLTkwNzYtYzIwZjljYmUxMDk1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDEzMDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyYjAzOTA3ZjFmZDRjN2Q4OTU2MjlkOGEwYWNkNDA3NjJmNDBjMGRhYjAwZmRhZTVhOTJhMzgzYWYwMDRjYTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.z0mDgvB5ZsyAvF-QCHazH_ccr7V9KCdKejjPCf37ocg">
> * `Reject` button is always Red. I think we can make it lighter when it's not final reject similar to approving

Nice detail, I will implement it.

> * It would be good to change text from `Reject` to `Vote for rejection` but we don't have space for it :(. Update: we have that button inside large stage reject modal. Maybe it's better to add it also for grants/milestones

The rejection system is only now being implemented for USDC grants. 

> * Text: "Clicking Reject will change the status of this grant to Rejected". It's not right. Status will change after final reject but not after click. As I remember purpose of that text was to warn about one-click rejection. Since reject behavior is changed for large grants probably it's better just to make text as for approval, something like "Are you sure you want to vote for the rejection of this stage?". Update: it already fixed but only for large stages

The rejection system is only now being implemented for USDC grants. 

> * Not related to this PR but we have this text also for regular (eth) **stages**: "Clicking Reject will change the status of this grant to Rejected". Need to change "grant" to "stage"
> 
> > When you vote for rejection, the "Already rejected" label makes the button section look a bit weird, changing to "Already voted" makes it consistant with approvals and it looks nicely styled:
> 
> note: for me, it's still weird 🤷‍♂️ . Mac 16 inches, browser window is full monitor width, scaling is 100%. <img alt="image" width="590" src="https://private-user-images.githubusercontent.com/25638585/434702824-3e56ba86-1c8e-42f3-8f8b-944ebdb93360.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDA5NjEsIm5iZiI6MTc0NTI0MDY2MSwicGF0aCI6Ii8yNTYzODU4NS80MzQ3MDI4MjQtM2U1NmJhODYtMWM4ZS00MmYzLThmOGItOTQ0ZWJkYjkzMzYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDEzMDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4ZGRhZmJhMjlkZTdkMWFlZDg3ZWYxNGU5ZjRjOWQ3OGQwN2Y0N2Y2NzU2NjYwNTQ1OWRmNzM4NGYzODBkZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LAjt-DoTAumkfcVx4KyKchDUfc13G5GvL34S6tXDMko">

:-( What resolution are you using?

rin-st commented:
> > * It would be good to change text from `Reject` to `Vote for rejection` but we don't have space for it :(. Update: we have that button inside large stage reject modal. Maybe it's better to add it also for grants/milestones
> 
> The rejection system is only now being implemented for USDC grants.

Yes I know, I wrote about large grants/large milestones


> > * Text: "Clicking Reject will change the status of this grant to Rejected". It's not right. Status will change after final reject but not after click. As I remember purpose of that text was to warn about one-click rejection. Since reject behavior is changed for large grants probably it's better just to make text as for approval, something like "Are you sure you want to vote for the rejection of this stage?". Update: it already fixed but only for large stages
> 
> The rejection system is only now being implemented for USDC grants.

Same, please check large milestones and large grants

> :-( What resolution are you using?

<img width="480" alt="image" src="https://github.com/user-attachments/assets/feb8b0a7-e979-4462-8d5d-487dde8ac2bb" />

But it's retina so 2 times less. It should be the same as 1728x1117

damianmarti commented:
> > > * It would be good to change text from `Reject` to `Vote for rejection` but we don't have space for it :(. Update: we have that button inside large stage reject modal. Maybe it's better to add it also for grants/milestones
> > 
> > 
> > The rejection system is only now being implemented for USDC grants.
> 
> Yes I know, I wrote about large grants/large milestones

Sorry, I don't understand. The reject button on large grant and large stage say `Vote for rejection` on both. On milestones, it says "Reject" because only one reject is needed for milestones right now.


> 
> > > * Text: "Clicking Reject will change the status of this grant to Rejected". It's not right. Status will change after final reject but not after click. As I remember purpose of that text was to warn about one-click rejection. Since reject behavior is changed for large grants probably it's better just to make text as for approval, something like "Are you sure you want to vote for the rejection of this stage?". Update: it already fixed but only for large stages
> > 
> > 
> > The rejection system is only now being implemented for USDC grants.
> 
> Same, please check large milestones and large stages

Sorry, I don't understand. The second reject is the final reject, so it's ok to show this message there. Or maybe I'm understanding something wrong.

> 
> > :-( What resolution are you using?
> 
> <img alt="image" width="480" src="https://private-user-images.githubusercontent.com/25638585/435707300-feb8b0a7-e979-4462-8d5d-487dde8ac2bb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDgwMzUsIm5iZiI6MTc0NTI0NzczNSwicGF0aCI6Ii8yNTYzODU4NS80MzU3MDczMDAtZmViOGIwYTctZTk3OS00NDYyLThkNWQtNDg3ZGRlOGFjMmJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE1MDIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwYjI0YTZkYTBlMGE4ZWY3NmQ5ZWM3NGEwMzU1YmRiYmVlMDAxODZlNGRjYjljMzg3Y2QyZjE5ZTc2NmIyYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZTQive1t0YDA3L3S9nmCFok-e25JQ79q3OvyoGKiIsg">
> But it's retina so 2 times less. It should be the same as 1728x1117. Not very critical, but high chance that admins use macs

Thanks!

rin-st commented:
> Sorry, I don't understand. The reject button on large grant and large stage say Vote for rejection on both. On milestones, it says "Reject" because only one reject is needed for milestones right now.

> Sorry, I don't understand. The second reject is the final reject, so it's ok to show this message there. Or maybe I'm understanding something wrong.

Sorry Damu! My bad, it works as it should. But I'm sure when I wrote the comment first time I checked large grants. Maybe it was some wrong grants/stages/milestones or I don't know. I couldnt missclick so much times

damianmarti commented:
> > Sorry, I don't understand. The reject button on large grant and large stage say Vote for rejection on both. On milestones, it says "Reject" because only one reject is needed for milestones right now.
> 
> > Sorry, I don't understand. The second reject is the final reject, so it's ok to show this message there. Or maybe I'm understanding something wrong.
> 
> Sorry Damu! My bad, it works as it should. But I'm sure when I wrote the comment first time I checked large grants. Maybe it was some wrong grants/stages/milestones or I don't know. I couldnt missclick so much times

No problem! Thanks!

damianmarti commented on packages/nextjs/app/admin/_components/LargeGrantProposal.tsx:
```diff
@@ -70,7 +75,11 @@ export const LargeGrantProposal = ({ proposal, userSubmissionsAmount, isGrant }:
-      <div className="px-5 py-8 bg-gray-100">
+      <div
+        className={`px-5 pt-8 pb-${
+          latestStage.approvalVotes.length > 0 || latestStage.rejectVotes.length > 0 ? "2" : "8"
```
Changed!

damianmarti commented:
> Looks good, thanks! Some comments/nitpicks
> 
> Bug: I can vote for both rejection and approval from the same account. It should be restricted
> 
> ui:
> 
> * We have `Cancel` button for large stages and large grants but don't have for milestones and for regular grants/stages. Probably it's easier just remove that button at all
> * Modal buttons has different sizes and paddings across all page modals. For example, `Cancel` button is smaller than `Reject`. 48px vs 52px height. I think they should be the same. Probably other PR

Created issue for this #85 

> 
> <img alt="image" width="558" src="https://private-user-images.githubusercontent.com/25638585/434687716-7ef5e14b-934e-4857-b5a5-bb41f40cce92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNTQ2NjYsIm5iZiI6MTc0NTI1NDM2NiwicGF0aCI6Ii8yNTYzODU4NS80MzQ2ODc3MTYtN2VmNWUxNGItOTM0ZS00ODU3LWI1YTUtYmI0MWY0MGNjZTkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE2NTI0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhYWI0MWE3NTg1MDczNzZkMzEyMzRjYjZiODY4OWFmOTM5ZGY0MTYyNDM4MjBiZmI5M2UyMzkzMjdmNTg4YzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GQiM2F8absfR07lRQNrE1ku_ybxNwcHz_xCc0GIPZk0"> <img alt="Screenshot 2025-04-17 at 10 31 19" width="491" src="https://private-user-images.githubusercontent.com/25638585/434693721-ab6cb6f2-218c-4e6c-9076-c20f9cbe1095.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNTQ2NjYsIm5iZiI6MTc0NTI1NDM2NiwicGF0aCI6Ii8yNTYzODU4NS80MzQ2OTM3MjEtYWI2Y2I2ZjItMjE4Yy00ZTZjLTkwNzYtYzIwZjljYmUxMDk1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE2NTI0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4OTkxMmE4MDM3YWYxMzY5ZDJiYzEwNWVlMWFkOTc3NDBhODA1ZmRkNGU2YzIwYWY3OTg2YTU2ZTY0MGVjODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ot9VZtyp9N8KK3I51ZFwQ93TlQlImD77WAGtH1iwcrk">
> * `Reject` button is always Red. I think we can make it lighter when it's not final reject similar to approving

Final reject uses red button variant, and vote for rejection uses red-secondary now.

> * It would be good to change text from `Reject` to `Vote for rejection` but we don't have space for it :(. Update: we have that button inside large stage reject modal. Maybe it's better to add it also for grants/milestones
> * Text: "Clicking Reject will change the status of this grant to Rejected". It's not right. Status will change after final reject but not after click. As I remember purpose of that text was to warn about one-click rejection. Since reject behavior is changed for large grants probably it's better just to make text as for approval, something like "Are you sure you want to vote for the rejection of this stage?". Update: it already fixed but only for large stages

I will fix it when I implement the rejection system for ETH grants.

> * Not related to this PR but we have this text also for regular (eth) **stages**: "Clicking Reject will change the status of this grant to Rejected". Need to change "grant" to "stage"
> 
> > When you vote for rejection, the "Already rejected" label makes the button section look a bit weird, changing to "Already voted" makes it consistant with approvals and it looks nicely styled:
> 
> note: for me, it's still weird 🤷‍♂️ . Mac 16 inches, browser window is full monitor width, scaling is 100%. <img alt="image" width="590" src="https://private-user-images.githubusercontent.com/25638585/434702824-3e56ba86-1c8e-42f3-8f8b-944ebdb93360.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNTQ2NjYsIm5iZiI6MTc0NTI1NDM2NiwicGF0aCI6Ii8yNTYzODU4NS80MzQ3MDI4MjQtM2U1NmJhODYtMWM4ZS00MmYzLThmOGItOTQ0ZWJkYjkzMzYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE2NTI0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwZjhjNGU5MWEyNjVmOTJiMjBkNDUzM2IwYTllZTdhYjc4NWFhZmE3NTU5MjgxZTY3Y2E5ZWFiYmM3NTc4M2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OXAJcRbkggpAqHiLq5HCRdRoY4rPKn7tY23IXnEuneg">

Fixed in a quick way, changing **Already voted** to **Voted**.

rin-st commented:
Thanks Damu! One more nit: 

> Final reject uses red button variant, and vote for rejection uses red-secondary now.

<img width="1133" alt="Screenshot 2025-04-22 at 16 56 51" src="https://github.com/user-attachments/assets/d216e298-4988-47ca-9c36-137e44a1f19c" />


For regular grants and stages we have light-red button on card but red button on modal. But since we have warning message inside that modal it's not so critical

Feel free to ignore it. Other than that, lgtm!

damianmarti commented:
> Thanks Damu! One more nit:
> 
> > Final reject uses red button variant, and vote for rejection uses red-secondary now.
> 
> <img alt="Screenshot 2025-04-22 at 16 56 51" width="1133" src="https://private-user-images.githubusercontent.com/25638585/436108161-d216e298-4988-47ca-9c36-137e44a1f19c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzMzI1MjQsIm5iZiI6MTc0NTMzMjIyNCwicGF0aCI6Ii8yNTYzODU4NS80MzYxMDgxNjEtZDIxNmUyOTgtNDk4OC00N2NhLTljMzYtMTM3ZTQ0YTFmMTljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIyVDE0MzAyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkMWZhYjQyMWJlOGZjZmQ4MGZkZjM4NGU5ZDdkZjdiNTJmNjVmOGU2YmRkZmRlNDUxZjBmMDBiMWI3ODYwN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.n7X3e-v3sdY9Zt1x3rB6j0EsY3T6RutyK5GIYj8Di2Q">
> For regular grants and stages we have light-red button on card but red button on modal. But since we have warning message inside that modal it's not so critical
> 
> Feel free to ignore it. Other than that, lgtm!

Thanks!!

I will check this when I implement the rejection system for ETH grants.
